### ISSUE TITLE
Service Fabric Stateful Service consuming large amount of unmanaged memory

### ISSUE THREAD

**Issue by johnkattenhorn** [Azure/service-fabric-issues#695]
This is a duplicate of a post I made on Stackflow here:

(https://stackoverflow.com/questions/47681701/service-fabric-stateful-service-consuming-large-amounts-of-unmanaged-memory )

I have a Stateful Service which consumes increasing amount of memory until the service is restarted or process is killed in which then releases the memory.

See the diagram below; it looks like the typical saw-tooth type problem of memory leaking somewhere. 

<img width="399" alt="image" src="https://user-images.githubusercontent.com/616249/33681453-7dd29562-dabc-11e7-9625-faab1163a7be.png">

We used DotMemory to run some analysis of the memory usage of a single node within the cluster and it reported the vast majority of the memory being consumed was in unmanaged memory.

<img width="363" alt="image" src="https://user-images.githubusercontent.com/616249/33681464-8830ca7e-dabc-11e7-956e-efd6fd8349cb.png">

Just before we cycled the stateful service we took a memory dump file to see if we could learn anything further using WinDbg.

I'm no WinDbg expert but I followed this article which seemed to suggest that most of the memory was being consumed by Heap Stack (http://hacksoflife.blogspot.co.uk/2009/06/heap-debugging-memoryresource-leak-with.html) 

It suggested that I should use some additional commands to get a stack trace but I didn't do this prior to taking the dmp file (gflags.exe /i yourApplication.exe +ust).

Is there anyone who could help me diagnose the issue using the dmp file I have ? 

Could someone validate that the steps mentioned in the article would be worth following to try to find this issue ?#

Has anyone experience this kind of issue with Stateful Services before ?

**Comment by MedAnd** [comment]
Are you using server or desktop GC? We’ve noticed similar when using server GC as per #320

**Comment by MedAnd** [comment]
Back in early versions we also had memory leaks in proxy creation logic, I think we used a now deprecated way of creating a proxy factory... would recommend using the latest version of Service Fabric and SDK, maybe check proxy creation logic looks something like below... also try testing if not creating FabricClient in the constructor helps, rather create per call?

```
IMyActor proxy = ActorProxy.Create<IMyActor>(ActorId.CreateRandom(), new Uri(serviceUri));
await proxy.StartProcessingAsync(CancellationToken.None);
```

**Comment by johnkattenhorn** [comment]
With the help of the support department we found an issue with FabricClient. Apparently there's a known issue with disposing of the FabricClient and is due to be fixed in SDK 6.2.

For now we've migrated our code to use a static variable to hold a single instance of FabricClient per service.
